rmw_zenoh_cpp/src/detail/rmw_client_data.cpp

```diff
@@ -0,0 +1,572 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_client_data.hpp"
+
+#include <fastcdr/FastBuffer.h>
+
+#include <cinttypes>
+#include <limits>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <utility>
+
+#include "attachment_helpers.hpp"
+#include "cdr.hpp"
+#include "rmw_context_impl_s.hpp"
+#include "message_type_support.hpp"
+#include "logging_macros.hpp"
+#include "qos.hpp"
+
+#include "rcpputils/scope_exit.hpp"
+
+#include "rmw/error_handling.h"
+#include "rmw/get_topic_endpoint_info.h"
+#include "rmw/impl/cpp/macros.hpp"
+
+namespace rmw_zenoh_cpp
+{
+// A static mutex whose lifetime will extend beyond that of the internal mutex_.
+// This is needed as the decrement_queries_in_flight callback that me be executed
+// after ClientData is deallocated will try to lock an invalid mutex and thus lead
+// to UB. This only happens in rclcpp_action/test_client.
+static std::mutex num_in_flight_mutex;
```

Comment:
I'm not sure I understand how this fixes the problem, and I'm not sure it is required anymore.  If `client_data` has gone out of scope, then calling `client_data->decrement_queries_in_flight()` is already UB, never mind accessing the `std::mutex` within it (it will usually happen to work, but that's besides the point).

But we really shouldn't get into this situation anymore anyway, because `NodeData::delete_client_data` will only be deleting the structure if there is nothing in flight.  (The potential memory leak is worse here, in that `client_data_drop` isn't cleaning up anymore if this is the "last" thing in flight, but that is not UB).

In other words, I don't see why we can't embed this mutex in the `ClientData` structure.  But there may be something I'm missing.

Reply:
I don't think I implemented this correctly.

My approach here was to avoid UBs only since memory leaks should never happen given the use of shared_ptrs.

If `client_data` is deallocated by resetting an `rclcpp::Node` ptr (as done in one of the tests), dereferencing `client_data` within `client_data_drop` like `client_data-> decrement_queries_in_flight()` would be UB. In this scenario, even though we do not remove the `client_data` from `NodeData::clients_` when `NodeData::delete_client_data()` is called, it will get removed when the `NodeData`  is deleted by the reset call. 

Ideally we have a static map of ClientData rawptr address to num_in_flight. Then `client_data_drop` would lookup this map and decrement the count. We would increment the count in the same map.

But I don't know if it worth the effort. Please let me know what you think.

Reply:
Alright I think I have a better solution implemented in https://github.com/ros2/rmw_zenoh/pull/293/commits/51f2917e5eae64937ddf0643145e4e69e9f424c3.